Fix Bug 1205858, adds new default contact us tab to contact page

schalkneethling opened the pull request:
@jpetto r?

@flodolo I am pretty sure this is not localized but, just putting an l10n label on this in case. If indeed it is not needed, you can just remove the label. Thanks!

flodolo commented:
Confirming, not localized.

jpetto commented on bedrock/mozorg/templates/mozorg/contact/contact-landing.html:
```diff
@@ -0,0 +1,66 @@
+        <li><a href="https://wiki.mozilla.org/People:MozSpaces_Guidelines">{{ _('I want to hold an event in a Mozilla space')}}</a></li>
+        <li><a href="https://bugzilla.mozilla.org/form.dev-engagement-event">{{ _('I want Mozilla to sponsor my event')}}</a></li>
+        <li><a href="https://www.mozilla.org/styleguide/identity/mozilla/branding/">{{ _('I have a question about using Mozilla logos')}}</a></li>
+        <li><a href="#press_enquiry" id="press_enquiry_toggle">{{ _('I’m a member of the Press and have a question for Mozilla') }}</a></li>
```
Should this `<li>` be removed until the press contact form is completed ([bug 1210426](https://bugzilla.mozilla.org/show_bug.cgi?id=1210426))?

schalkneethling replied:
Indeed, completely neglected to remove it.

jpetto commented on bedrock/mozorg/templates/mozorg/contact/contact-landing.html:
```diff
@@ -0,0 +1,66 @@
+      <ul>
+        <li><a href="https://support.mozilla.org/">{{ _('I’m having problems with using Firefox')}}</a></li>
+        <li><a href="https://sendto.mozilla.org/page/contribute/givenow-seq?preset=2&ref=Donation_Form&utm_campaign=Donation_Form&utm_source=mozilla.org&utm_medium=referral&utm_content=contact_us_page">{{ _('I want to donate to Mozilla')}}</a></li>
+        <li><a href="https://www.mozilla.org/foundation/trademarks/policy/">{{ _('I have questions about copyright trademarks and Mozilla')}}</a></li>
```
Use the `url` helper here?

jpetto commented on bedrock/mozorg/templates/mozorg/contact/contact-landing.html:
```diff
@@ -0,0 +1,66 @@
+        <li><a href="mailto:research_requests@mozilla.com">{{ _('I need information for an academic paper')}}</a></li>
+        <li><a href="https://wiki.mozilla.org/People:MozSpaces_Guidelines">{{ _('I want to hold an event in a Mozilla space')}}</a></li>
+        <li><a href="https://bugzilla.mozilla.org/form.dev-engagement-event">{{ _('I want Mozilla to sponsor my event')}}</a></li>
+        <li><a href="https://www.mozilla.org/styleguide/identity/mozilla/branding/">{{ _('I have a question about using Mozilla logos')}}</a></li>
```
And here?

jpetto commented on bedrock/mozorg/templates/mozorg/contact/contact-landing.html:
```diff
@@ -0,0 +1,66 @@
+        <h4>{{ _('Find us online') }}</h4>
+        <ul>
+          <li><a href="https://facebook.com/mozilla" hreflang="en-US">{{ _('Facebook') }}</a></li>
+          <li><a href="https://twitter.com/mozilla" hreflang="en-US">{{ _('Twitter') }}</a></li>
```
Put a `rel="external"` on these two social links?

jpetto commented on bedrock/mozorg/templates/mozorg/contact/contact-landing.html:
```diff
@@ -0,0 +1,66 @@
+        <ul>
+          <li><a href="https://facebook.com/mozilla" hreflang="en-US">{{ _('Facebook') }}</a></li>
+          <li><a href="https://twitter.com/mozilla" hreflang="en-US">{{ _('Twitter') }}</a></li>
+          <li><a href="/about/forums/">{{ _('Forums &amp; e-mail lists') }}</a></li>
```
Use `url` helper here?

schalkneethling replied:
Yeah, good point. This was all copied from preexisting code but, no reason not to improve it ;)

jpetto commented on media/js/mozorg/contact-spaces.js:
```diff
@@ -115,12 +115,16 @@
-            $('.nav-category li[data-id="' + id + '"]').addClass('current');
+            if (tab === 'spaces') {
+                // set the current list menu navigation item for the active space
+                $('.nav-category li[data-id="' + id + '"]').addClass('current');
```
I believe this line is still needed for both `spaces` and `communities`, as both have side navs.

jpetto replied:
Still need to remove the conditional around this line?

jpetto replied:
Ah, I see it's repeated below. Would be easier/more DRY to just have the line once, not wrapped in a conditional though, yeah?

schalkneethling replied:
:+1:

jpetto commented on media/js/mozorg/contact-spaces.js:
```diff
@@ -115,12 +115,16 @@
-            mozMap.toggleCommunitySubMenu();
+            if (tab === 'communities') {
+                // hide the community sub menu's
+                $('.accordion .submenu').hide();
```
I know the logic isn't changed in this PR, but I feel this line would be clearer if targeted at the `nav-communities` ID, as it appears the intent is to affect a specific element and not all `.accordion` elements. Thoughts?

Also, looks like a conditional around these two `communities` related lines should have been there before this PR, as they shouldn't need to fire for `spaces` or `contact`. :+1:

jpetto commented on media/js/mozorg/contact-spaces.js:
```diff
@@ -778,6 +805,30 @@
+         */
+        showContactUsContent: function(url, cacheId) {
+            var $current = $('.category-tabs li.current');
+            var id = $current.data('id');
```
Doesn't look like this local variable is used.

jpetto commented on media/js/mozorg/contact-spaces.js:
```diff
@@ -778,6 +805,30 @@
+        showContactUsContent: function(url, cacheId) {
+            var $current = $('.category-tabs li.current');
+            var id = $current.data('id');
+            var contentUrl = url || current.prop('href');
```
Missing `$` in front of `current`.

jpetto commented on media/css/mozorg/contact-spaces.less:
```diff
@@ -437,6 +437,11 @@
 
+/* override float rule for contact us tab */
+.entry.contact-us {
+    float: none;
```
Should include the `.clearfix()` mixin here, as contains floated content?

schalkneethling replied:
Did not run into any floating issues without the mixin so, perhaps it is not required?

jpetto replied:
The `.clearfix` on the parent `.container` is saving us from layout issues. If we had content underneath "Questions & feedback" however, there would be issues. I tend to err on the side of clearing floats. Doesn't hurt anything. :wink:

alexgibson commented:
Before merging this it might be a good idea to double-check mcom-tests to see if anything needs updating :+1:

jpetto commented:
Looks like there's a bug on mobile, as the select boxes for both Spaces and Communities are visible when toggling back to the "Contact" tab.

![screen shot 2015-10-08 at 1 39 32 pm](https://cloud.githubusercontent.com/assets/317532/10376202/36da5b54-6dc2-11e5-99dc-b57e0e26d02b.png)

Also, desktop & tablet views looks a *little* bit off due to the amount of white space on the right side of the content. Since there's no side nav, perhaps we should remove the width restriction for the Contact content?

schalkneethling commented:
@jpetto addressed all comments. @alexgibson so, this one will be blocked until the mcom tests are updated?

This, for example, will need to be updated:
https://github.com/mozilla/mcom-tests/blob/master/tests/test_contact.py#L29

alexgibson commented:
I think I might have spotted a history bug in this change, but will double-check soon to see if I can reproduce (currently mobile). Yes I think we'll likely need to update mcom-tests for this change too.

jpetto commented:
Hm, found another layout issue - tabs are unevenly spaced at mobile wide:

![screen shot 2015-10-08 at 3 37 25 pm](https://cloud.githubusercontent.com/assets/317532/10379160/a114aef6-6dd2-11e5-9166-fb4b5805eaef.png)

Could probably get rid of that top border for Contact mobile view as well, as there's no select box up top to separate.

jpetto commented on media/js/mozorg/contact-spaces.js:
```diff
@@ -115,12 +115,17 @@
-            mozMap.toggleCommunitySubMenu();
+            if (tab === 'communities') {
+                // hide the community sub menu's
+                $('#nav-communities').hide();
```
Ah, but this still has to target the child `.submenu`'s. Should have been more clear in my previous comment.

jpetto replied:
Just a reminder to switch this to `$('#nav-communities .submenu').hide();` in the next update.

jpetto commented:
Jeez, just noticed the tabs are unevenly spaced at desktop size as well. So subtle (and noisy with map bg) that I didn't notice it at first.

![screen shot 2015-10-08 at 4 13 55 pm](https://cloud.githubusercontent.com/assets/317532/10380118/9fa3e208-6dd7-11e5-8976-aa41a212dee2.png)

jpetto commented on media/js/mozorg/contact-spaces.js:
```diff
@@ -311,6 +322,8 @@
                     $('#nav-communities, #meta-communities').show();
                 });
+            } else if (tab === 'contact') {
+                $('#nav-spaces, #meta-spaces, #nav-communities, #meta-communities').fadeOut(100);
```
For consistency, let's keep the logic uniform and make the condition here `(tab === 'contact' && wideMode)`. Then we can simplify/ the `toggleMobileNav` function below to be:

```
} else if (tab === 'contact' && !wideMode) {
    $('#nav-communities-select, #meta-communities, #nav-spaces-select, #meta-spaces').hide();
}
```

jpetto commented:
I'm seeing a history bug - possibly the one @alexgibson noticed. Load `/contact/`, click the "Spaces" tab, wait for content to load, click back button. The content does not switch back to "Contact". If you go from "Spaces" to "Communities" (or vice versa), the content updates correctly using the back/forward buttons.

alexgibson commented:
> I'm seeing a history bug - possibly the one @alexgibson noticed. Load /contact/, click the "Spaces" tab, wait for content to load, click back button. The content does not switch back to "Contact". If you go from "Spaces" to "Communities" (or vice versa), the content updates correctly using the back/forward buttons.

Yep this is what I'm seeing too. Specifically, I think the new code needs to handle [this condition](https://github.com/mozilla/bedrock/blob/master/media/js/mozorg/contact-spaces.js#L160-L161)

schalkneethling commented:
Ok, PR opened against mcom tests https://github.com/mozilla/mcom-tests/pull/443

alexgibson commented:
I did a bit of debugging locally on the history bug and I think these areas may still need updating:

https://github.com/mozilla/bedrock/blob/master/media/js/mozorg/contact-spaces.js#L534-L538
https://github.com/mozilla/bedrock/blob/master/media/js/mozorg/contact-spaces.js#L395

Also, we have a much [simplified version](https://github.com/mozilla/bedrock/blob/master/media/js/mozorg/contact-spaces-ie7.js) for older IE browsers. Does this need updating at all for this new tab?

schalkneethling commented:
@jpetto test all the things but, I need to update the IE bits that @alexgibson mentioned above. Was not even aware of that. Thanks Alex

jpetto commented:
The history bug seems to be fixed. :+1: 

However, during testing, I noticed something odd. On any initial page load (`/contact/`, `/contact/spaces/`, `/contact/communities/`), the JS wipes out the main content, then an XHR request is made requesting the exact same page, which re-populates the main content. Essentially, it looks like the mechanism for smoothly swapping content when clicking tabs is operating at page load as well, resulting in a totally unnecessary request/performance hit.

Check the network tab in dev tools and you'll see an XHR request to the same page initially loaded. :confounded: 

I believe these three functions are all culprits:

`showSpace`
`showCommunityContent`
`showContactUsContent`

This behavior doesn't look to be introduced with this PR, but it sure would be nice to fix it up soon. Could be in this PR, or file a new bug if other matters are pressing.

alexgibson commented:
@jpetto - nice spot. Not sure when this crept in, but you're right we should fix it. Probably just need to add a check to make sure the ajax doesn't get called on page load :+1:

schalkneethling commented:
Thanks @jpetto I will resolve that in this PR as well.

schalkneethling commented:
@jpetto updated. Let me know if there are other items you run into. Thanks!

jpetto commented:
The last thing I see is that the initial content is not cached, which can cause an extra XHR request. If I load `/contact/` and click the "Spaces" tab, an XHR request is made to get the "Spaces" content. All good there. If I then click the back button or the "Contact" tab, another XHR request is made to get the "Contact" content (even though that content was retrieved on initial page load).

Maybe we could factor out the [caching mechanism from the AJAX request](https://github.com/mozilla/bedrock/blob/master/media/js/mozorg/contact-spaces.js#L1078-L1084) and use it in the [`setInitialContentState`](https://github.com/mozilla/bedrock/blob/master/media/js/mozorg/contact-spaces.js#L395) function to cache initially loaded content?

schalkneethling commented:
@jpetto So, it seems the cacheid is not being set, let me look into it. Will update the PR and let you know. Thanks!

schalkneethling commented:
@jpetto yes, so as you mentioned and I just confirmed, the cacheId is not set and hence the ajax request happens. Now the problem is, on initial load there are no ajax requests as the appropriate page is loaded directly.

I guess the way to get this to work is to change the way the page works overall in that the content is always loaded via Ajax, even on initial load i.e. the same page is loaded for `/contact/`, `/contact/spaces/` and 'contact/communities/` but, based on the URL, the appropriate content is loaded via Ajax thus, setting the cacheId from initial load.

This should probably be part of the refactor that is on the cards for the very near future. Perhaps there is a stop gap I am overlooking that we can implement in this bug but else, it is probably out of scope for this bug. Thoughts? Thanks!

jpetto commented:
I *think* the fix could be in factoring out the caching mechanism from the AJAX request, then using it on page load *and* AJAX requests. From what I've seen of the code, AJAX isn't required to populate the cache, but they are tightly coupled at the moment.

However, I feel the page is at a good state in this PR. The redundant XHR request at page load has been removed, which is a big win (and was also outside the scope of this bug). In the interest of getting this new tab out the door, I'd be fine with opening another bug to finish up polishing the caching performance.

Just needs a squash and we'll be good to go. r+! :rocket:

jpetto commented:
Also, just noting that [this pre-existing rewrite rule](https://github.com/mozilla/bedrock/blob/master/etc/httpd/global.conf#L298) should handle this new page:

`RewriteRule ^/(\w{2,3}(?:-\w{2})?/)?contact(.*)$ /b/$1contact$2 [PT]`

schalkneethling commented:
Awesome, thx for the review @jpetto Squashed. With regards to the rewrite rule yes, it will work just fine. There was a redirect that would redirect requests to `spaces/` but is removed here.

So, let's merge this first. I will then merge master onto the inquiry page, add the link and we can merge that one.
